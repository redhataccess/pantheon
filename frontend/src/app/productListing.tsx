import React, { Component } from 'react';
import {
  DataList, DataListItem, DataListCell, DataListItemRow, DataListItemCells, DataListAction, FormGroup,
  OptionsMenu, OptionsMenuItem, OptionsMenuToggle, TextInput
} from '@patternfly/react-core';
import '@app/app.css';
import { ProductDetails } from '@app/productDetails';
import { Redirect } from 'react-router-dom'

class ProductListing extends Component<any, any, any> {

  constructor(props) {
    super(props);
    this.state = {
      allProducts: [],
      initialLoad: true,
      input: '',
      isEmptyResults: false,
      isOpen: false,
      isProductDetails: false,
      loggedinStatus: false,
      login: false,
      productName: '',
      redirect: false,
      results: []
    };
  }

  public componentDidMount() {

    if (!this.state.loggedinStatus && this.state.initialLoad === true) {
      fetch("/system/sling/info.sessionInfo.json")
        .then(response => response.json())
        .then(responseJSON => {
          if (responseJSON.userID !== 'anonymous') {
            this.setState({ loggedinStatus: true })
          }
        })
    }

    if (this.props.match !== undefined) {

      // prop will be true if it comes through nav links
      if (this.props.match.isExact === true) {
        this.state.results.map(data => {
          (data.isOpen as any) = false
        });
        this.setState({ isProductDetails: false })
      }

      // setting prop to false once it comes through nav links
      this.props.match.isExact = false;

    }

    this.checkAuth()
    this.getProducts(this.state.allProducts)
  }

  // render method transforms the react components into DOM nodes for the browser.
  public render() {


    return (
      <React.Fragment>
        {this.state.isProductDetails && (<ProductDetails productName={this.state.productName} />)}
        {!this.state.isProductDetails && (
          <div>
            <FormGroup
              label="Search Products"
              fieldId="search"
            >
              <div className="row-view">
                <TextInput id="search" type="text" onChange={this.setInput} placeholder="Type product name to search" value={this.state.input} />
              </div>
            </FormGroup>
            <DataList aria-label="single action data list example ">
              {!this.state.isEmptyResults && (
                <DataListItem aria-labelledby="single-action-item1">
                  <DataListItemRow>
                    <DataListItemCells
                      dataListCells={[
                        <DataListCell key="primary content">
                          <span className="sp-prop-nosort" id="product-name">Product Name</span>
                        </DataListCell>,
                        <DataListCell key="secondary content" width={2}>
                          <span className="sp-prop-nosort" id="product-description">Product Description</span>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>
              )}

              {!this.state.isEmptyResults && this.state.results.map(data => (
                <DataListItem aria-labelledby="multi-actions-item1">
                  <DataListItemRow>
                    <DataListItemCells key={data["jcr:uuid"]}
                      dataListCells={[
                        <DataListCell key="primary content">
                          <span id="{data.name}">{data.name}</span>
                        </DataListCell>,
                        <DataListCell key="secondary content" width={2}>{data.description}</DataListCell>,
                        <DataListCell key="Dropdown content">
                          <DataListAction
                            aria-labelledby="multi-actions-item1 {data['jcr:uuid']}"
                            id="{data['jcr:uuid']}"
                            aria-label="Actions"
                          >
                            <OptionsMenu
                              isPlain={true}
                              id={data['jcr:uuid']}
                              menuItems={[
                                <OptionsMenuItem onSelect={this.onSelect(event, data)} key="dropdown">Product Details</OptionsMenuItem>]}
                              isOpen={data.isOpen}
                              toggle={<OptionsMenuToggle onToggle={this.onToggle(data['jcr:uuid'])} />} />
                          </DataListAction>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>))}
              {this.state.isEmptyResults && (
                <DataListItem aria-labelledby="single-action-item0" data-testid="emptyResults">
                  <DataListItemRow>
                    <DataListItemCells
                      dataListCells={[
                        <DataListCell key="primary content" width={2}>
                          <span className="sp-prop-nosort" id="product-name">No products found</span>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>
              )}
            </DataList>
          </div>)}
        <div>
          {this.loginRedirect()}
        </div>
      </React.Fragment>
    );
  }

  private getProducts = (allProducts) => {
    this.setState({ initialLoad: false })
    fetch(this.getProductsUrl())
      .then(response => response.json())
      .then(responseJSON => {
        let key; let singleProduct;
        for (const i of Object.keys(responseJSON.results)) {
          key = Object.keys(responseJSON.results)[i];
          singleProduct = responseJSON.results[key];
          singleProduct = Object.assign({ "isOpen": false }, singleProduct)
          allProducts.push(singleProduct)
        }
        this.setState({ results: allProducts })
      })
      .then(() => {
        // console.log(this.state.loggedinStatus)
        if (Object.keys(this.state.results).length === 0) {
          this.setState({
            isEmptyResults: true
          });
        } else {
          this.setState({
            isEmptyResults: false
          });
        }
      })
  }

  private getProductsUrl() {
    const backend = "/content/products.query.json?nodeType=pant:product&orderby=name"
    return backend
  }

  private onToggle = (id) => (event: any) => {
    this.state.results.map(data => {
      if (data['jcr:uuid'] === id) {
        (data.isOpen as any) = !data.isOpen
        this.setState({ isProductDetails: false })
      }
    });
  };

  private onSelect = (event, data) => () => {
    this.setState({
      initialLoad: false,
      isProductDetails: !this.state.isProductDetails,
      productName: data.name,
    });
  };

  private setInput = input => {
    const versions: string[] = [];
    let searchString = '';
    this.setState({ input })
    this.state.allProducts.map(data => {
      searchString = '' + data.name
      if (searchString.toLowerCase().includes(input.toLowerCase())) {
        versions.push(data)
      }
    });
    this.setState({ results: versions })
  };

  private loginRedirect = () => {
    if (this.state.login) {
      return <Redirect to='/login' />
    } else {
      return ""
    }
  }

  private checkAuth = () => {
    fetch("/system/sling/info.sessionInfo.json")
      .then(response => response.json())
      .then(responseJSON => {
        if (responseJSON.userID === 'anonymous') {
          this.setState({ login: true })
        }
      })
  }
}

export { ProductListing }