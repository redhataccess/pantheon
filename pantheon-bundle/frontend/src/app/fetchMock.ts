if (!self.fetch) {
    self.fetch = (input: RequestInfo, init?: RequestInit | undefined) => {
        const p = new Promise<Response>((resolve, reject) => {
            // just don't do anything. this exists so that our tests don't try and actually resolve fetches,
            // (which is exactly what isomorphic-fetch tried to do while we were importing it...)
            // and calling nothing here allows us to skip all of our fetch.then() and fetch.catch() code.
        })
        return p
    }
}