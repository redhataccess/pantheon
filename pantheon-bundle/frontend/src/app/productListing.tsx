import React, { Component } from 'react';
import {
  DataList, DataListItem, DataListCell, DataListItemRow, DataListItemCells, DataListAction, FormGroup,
  OptionsMenu, OptionsMenuItem, OptionsMenuToggle, TextInput
} from '@patternfly/react-core';
import '@app/app.css';
import { ProductDetails } from '@app/productDetails';
import { CaretDownIcon } from '@patternfly/react-icons';

class ProductListing extends Component<any, any, any> {

  constructor(props) {
    super(props);
    this.state = {
      allProducts: [],
      input: '',
      isEmptyResults: false,
      isOpen: false,
      isProductDetails: false,
      productName: '',
      redirect: false,
      results: []
    };
  }

  public componentDidMount() {
    if (this.props.match !== undefined) {

      // prop will be true if it comes through nav links
      if (this.props.match.isExact === true) {
        this.state.results.map(data => {
          (data.isOpen as any) = false
        });
        this.setState({ isProductDetails: false })
      }

      // setting prop to false once it comes through nav links
      this.props.match.isExact = false;

    }

    this.getProducts(this.state.allProducts)
  }

  // render method transforms the react components into DOM nodes for the browser.
  public render() {


    return (
      <React.Fragment>
        {this.state.isProductDetails && (<ProductDetails productName={this.state.productName} />)}
        {!this.state.isProductDetails && (
          <div>
            <FormGroup
              label="Search Products"
              fieldId="search"
            >
              <div className="row-view">
                <TextInput id="search" type="text" onChange={this.setInput} placeholder="Type product name to search" value={this.state.input} />
              </div>
            </FormGroup>
            <DataList aria-label="single action data list example ">
              {!this.state.isEmptyResults && (
                <DataListItem aria-labelledby="single-action-item1">
                  <DataListItemRow>
                    <DataListItemCells
                      dataListCells={[
                        <DataListCell key="primary content">
                          <span className="sp-prop-nosort" id="product-name">Product Name</span>
                        </DataListCell>,
                        <DataListCell key="secondary content" width={2}>
                          <span className="sp-prop-nosort" id="product-description">Product Description</span>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>
              )}

              {!this.state.isEmptyResults && this.state.results.map(data => (
                <DataListItem aria-labelledby="multi-actions-item1">
                  <DataListItemRow>
                    <DataListItemCells key={data["jcr:uuid"]}
                      dataListCells={[
                        <DataListCell key="primary content">
                          <span id="{data.name}">{data.name}</span>
                        </DataListCell>,
                        <DataListCell key="secondary content" width={2}>{data.description}</DataListCell>,
                        <DataListCell key="Dropdown content">
                          <DataListAction
                            aria-labelledby="multi-actions-item1 {data['jcr:uuid']}"
                            id="{data['jcr:uuid']}"
                            aria-label="Actions"
                          >
                            <OptionsMenu
                              isPlain={true}
                              id={data['jcr:uuid']}
                              menuItems={[
                                <OptionsMenuItem onSelect={this.onSelect(event, data)} key="dropdown">Product Details</OptionsMenuItem>]}
                              isOpen={data.isOpen}
                              toggle={<OptionsMenuToggle onToggle={this.onToggle(data['jcr:uuid'])} toggleTemplate={<CaretDownIcon aria-hidden="true" />} aria-label="Sort by" hideCaret={true} />} />
                          </DataListAction>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>))}
              {this.state.isEmptyResults && (
                <DataListItem aria-labelledby="single-action-item0" data-testid="emptyResults">
                  <DataListItemRow>
                    <DataListItemCells
                      dataListCells={[
                        <DataListCell key="primary content" width={2}>
                          <span className="sp-prop-nosort" id="product-name">No products found</span>
                        </DataListCell>
                      ]}
                    />
                  </DataListItemRow>
                </DataListItem>
              )}
            </DataList>
          </div>)}
      </React.Fragment>
    );
  }

  private getProducts = (allProducts) => {
    fetch(this.getProductsUrl())
      .then(response => response.json())
      .then(responseJSON => {
        let key; let singleProduct;
        for (const i of Object.keys(responseJSON.results)) {
          key = Object.keys(responseJSON.results)[i];
          singleProduct = responseJSON.results[key];
          singleProduct = Object.assign({ "isOpen": false }, singleProduct)
          allProducts.push(singleProduct)
        }
        this.setState({ results: allProducts })
      })
      .then(() => {
        if (Object.keys(this.state.results).length === 0) {
          this.setState({
            isEmptyResults: true
          });
        } else {
          this.setState({
            isEmptyResults: false
          });
        }
      })
  }

  private getProductsUrl() {
    const backend = "/content/products.query.json?nodeType=pant:product&orderby=name"
    return backend
  }

  private onToggle = (id) => (event: any) => {
    this.state.results.map(data => {
      if (data['jcr:uuid'] === id) {
        (data.isOpen as any) = !data.isOpen
        this.setState({ isProductDetails: false })
      }
    });
  };

  private onSelect = (event, data) => () => {
    this.setState({
      isProductDetails: !this.state.isProductDetails,
      productName: data.name,
    });
  };

  private setInput = input => {
    const versions: string[] = [];
    let searchString = '';
    this.setState({ input })
    this.state.allProducts.map(data => {
      searchString = '' + data.name
      if (searchString.toLowerCase().includes(input.toLowerCase())) {
        versions.push(data)
      }
    });
    this.setState({ results: versions })
  };
}

export { ProductListing }