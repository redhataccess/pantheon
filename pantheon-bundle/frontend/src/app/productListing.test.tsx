import React from 'react'
import { ProductListing }  from '@app/productListing'
import '@app/fetchMock'

import { mount, shallow } from 'enzyme'
import { DataList, DataListItem, DataListItemCells, DataListItemRow, FormGroup, TextInput } from '@patternfly/react-core'
import renderer from 'react-test-renderer'

 const props = {
   match: exact => true
 }

describe('ProductListing tests', () => {
  test('should render ProductListing component', () => {
    const view = shallow(<ProductListing />)
    expect(view).toMatchSnapshot()
  })

   it('should render a Data List', () => {
     const wrapper = mount(<ProductListing />)
     const dataList = wrapper.find(DataList)
     expect(dataList.exists()).toBe(true)
   })

   it('should render a DataListItem', () => {
    const wrapper = mount(<ProductListing />)
    const dataListItem = wrapper.find(DataListItem)
    expect(dataListItem.exists()).toBe(true)
  })

  it('should render a DataListItemCells Element', () => {
    const wrapper = mount(<ProductListing />)
    const dataListItemCells = wrapper.find(DataListItemCells)
    expect(dataListItemCells.exists()).toBe(true)
  })

  it('should render a DataListItemRow element', () => {
    const wrapper = mount(<ProductListing />)
    const dataListItemRow = wrapper.find(DataListItemRow)
    expect(dataListItemRow.exists()).toBe(true)
  })

   it('should render a form group', () => {
     const wrapper = mount(<ProductListing />)
     const formGroup = wrapper.find(FormGroup)
     expect(formGroup.exists()).toBe(true)
   })

   it('should render a text input', () => {
     const wrapper = mount(<ProductListing />)
     const textInput = wrapper.find(TextInput)
     expect(textInput.exists()).toBe(true)
   })

   it('test props', () => {
     const productListing = mount(<ProductListing {...props} />).matchesElement
     expect(productListing.length === 1)
   })

   
   it('test getProducts function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.getProducts([{"product1": "product1 name"}])).toMatchSnapshot()
  })

   it('test getProductsUrl function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.getProductsUrl("/content/products.query.json?nodeType=pant:product&orderby=name")).toMatchSnapshot()
  })
   
   it('test setInput function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.setInput("test input")).toMatchSnapshot()
  })

   it('test loginRedirect function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.loginRedirect).toMatchSnapshot()
  })
 
  it('test checkAuth function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.checkAuth).toMatchSnapshot()
  })

  it('test componentDidMount function', () => {
    const wrapper = renderer.create(<ProductListing />)
    const inst = wrapper.getInstance()
    expect(inst.componentDidMount).toMatchSnapshot()
  })
})
