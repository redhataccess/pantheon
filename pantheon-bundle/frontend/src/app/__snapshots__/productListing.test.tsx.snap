// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListing tests should render ProductListing component 1`] = `
<Fragment>
  <FormGroup
    fieldId="search"
    label="Search Products"
  >
    <div
      className="row-view"
    >
      <ForwardRef
        id="search"
        onChange={[Function]}
        placeholder="Type product name to search"
        type="text"
        value=""
      />
    </div>
  </FormGroup>
  <DataList
    aria-label="single action data list example "
  >
    <DataListItem
      aria-labelledby="single-action-item0"
      data-testid="emptyResults"
    >
      <DataListItemRow>
        <DataListItemCells
          dataListCells={
            Array [
              <DataListCell
                width={2}
              >
                <span
                  className="sp-prop-nosort"
                  id="product-name"
                >
                  No products found
                </span>
              </DataListCell>,
            ]
          }
        />
      </DataListItemRow>
    </DataListItem>
  </DataList>
</Fragment>
`;
